  <div class="alert alert-info">
  Привет, Евгения! Меня зовут Светлана и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе. Предлагаю общаться на «ты». Но если это не удобно - дай знать, и мы перейдем на «вы».

  <div class="alert alert-success">
  <b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
  </div>
  <div class="alert alert-warning">
  <b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
  </div>
  <div class="alert alert-danger">
  <b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
  </div>
  <div class="alert alert-info">
  <b>👂 Совет:</b> Какие-то дополнительные материалы
  </div>
  Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
  Мне будет легче отследить изменения, если ты выделишь свои комментарии:
  <div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
  Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
   </div>

# Анализ базы данных для крупного сервиса для чтения книг по подписке

**Дано:** база данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг.

**Цель:** сформулировать ценностное предложение для нового продукта.

**Описание данных:**

*Таблица **books** содержит данные о книгах*:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

*Таблица **authors** содержит данные об авторах*:
- author_id — идентификатор автора;
- author — имя автора.

*Таблица **publishers** содержит данные об издательствах*:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

*Таблица **ratings** содержит данные о пользовательских оценках книг*:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

*Таблица **reviews** содержит данные о пользовательских обзорах на книги*:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

**План работы:**

Шаг 1. Загрузка данных.

Шаг 2. Исследование таблиц.

*Исследуйте таблицы — выведите первые строки, посчитайте количество строк в каждой таблице.*

Шаг 3. Решение задач:

1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

*Сделайте по одному SQL-запросу для решения каждого задания.*

*Выведите результат каждого запроса в тетрадке.*

*Опишите выводы по каждой из решённых задач.*

## Загрузка данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследование таблиц

### Books

In [2]:
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Посчитаем количество строк в таблице.

In [3]:
query = '''SELECT COUNT(*) FROM books'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   1000

### Authors

In [4]:
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Посчитаем количество строк в таблице.

In [5]:
query = '''SELECT COUNT(*) FROM authors'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    636

### Publishers

In [6]:
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Посчитаем количество строк в таблице.

In [7]:
query = '''SELECT COUNT(*) FROM publishers'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    340

### Ratings 

In [8]:
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Посчитаем количество строк в таблице.

In [9]:
query = '''SELECT COUNT(*) FROM ratings'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   6456

### Reviews

In [10]:
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Посчитаем количество строк в таблице.

In [11]:
query = '''SELECT COUNT(*) FROM reviews'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   2793

 <div class="alert alert-success">
 <b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
 </div>

## Решение задач

### Количество книг, вышедших после 1 января 2000 года

In [12]:
query = '''
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-01-01'
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**Вывод:** после 1 января 2000 года вышло 819 книг.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Количество обзоров и средняя оценка каждой книги

In [13]:
query = '''
WITH
rw AS (SELECT book_id,
       COUNT(review_id) AS cnt_review
FROM reviews 
GROUP BY book_id),

rt AS (SELECT book_id,
       AVG(rating) AS avg_rating
FROM ratings 
GROUP BY book_id)

SELECT b.book_id,
       b.title,
       rw.cnt_review,
       rt.avg_rating
FROM books AS b 
LEFT OUTER JOIN rw ON b.book_id = rw.book_id
LEFT OUTER JOIN rt ON b.book_id = rt.book_id
ORDER BY cnt_review DESC, avg_rating DESC
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  cnt_review  \
0        672   The Cat in the Hat and Other Dr. Seuss Favorites         NaN   
1        191  Disney's Beauty and the Beast (A Little Golden...         NaN   
2        387                               Leonardo's Notebooks         NaN   
3        221                          Essential Tales and Poems         NaN   
4         83    Anne Rice's The Vampire Lestat: A Graphic Novel         NaN   
..       ...                                                ...         ...   
995      446                                Moo  Baa  La La La!         1.0   
996      186                     Death: The High Cost of Living         1.0   
997      318                                  History of Beauty         1.0   
998      943               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1         1.0   
999        2                 1 000 Places to See Before You Die         1.0   

     avg_rating  
0      5.000000  
1      4.000000  
2      4.000000  
3      4.000000  
4      3.666667  
..          ...  
995    3.000000  
996    3.000000  
997    2.500000  
998    2.500000  
999    2.500000  

[1000 rows x 4 columns]

 <div class="alert alert-danger">
<s> <b>😔 Необходимо исправить:</b> В итоговую таблицу должны войти все книги, даже если у каких-то книг нет обзоров или оценок</s>
 </div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Поправила.</div>

**Вывод:** у каждой книги из 1000 есть оценки, отзывы есть для 994 книг.

Максимальное количество отзывов на книгу - 7, минимальное - 0.

Максимальная средняя оценка книги - 5 баллов, минимальная - 2,5 балла.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [14]:
query = '''
SELECT p.publisher,
       COUNT(b.book_id) AS cnt_books
FROM books AS b 
LEFT OUTER JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY cnt_books DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  cnt_books
0  Penguin Books         42

**Вывод:** наибольшее количество книг толще 50 страниц выпустило издательство Penguin Books.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Автор с самой высокой средней оценкой книг среди книг с 50 и более оценками

In [15]:
query = '''
WITH
r AS (SELECT book_id,
       COUNT(rating) AS cnt_rating,
       AVG(rating) AS avg_rating
FROM ratings 
GROUP BY book_id
HAVING COUNT(rating) >= 50
ORDER BY avg_rating DESC),

ba AS (SELECT b.book_id,
       a.author
FROM books AS b
LEFT OUTER JOIN authors as a ON b.author_id = a.author_id)

SELECT ba.author,
       AVG(r.avg_rating) AS avg_author_rating
FROM r 
LEFT OUTER JOIN ba ON r.book_id=ba.book_id
GROUP BY ba.author
ORDER BY avg_author_rating DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_author_rating
0  J.K. Rowling/Mary GrandPré           4.283844

**Вывод:** Автор с самой высокой средней оценкой книг среди книг с 50 и более оценками - J.K. Rowling/Mary GrandPré.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [16]:
query = '''
WITH
u AS (SELECT DISTINCT username,
       COUNT(rating_id) AS cnt_rating
FROM ratings 
GROUP BY DISTINCT username
HAVING COUNT(rating_id) > 48
ORDER BY cnt_rating DESC),

r AS (SELECT u.username,
       COUNT(rw.review_id) AS cnt_review
FROM u
LEFT OUTER JOIN reviews AS rw ON u.username = rw.username
GROUP BY u.username)

SELECT AVG(cnt_review)
FROM r
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

**Вывод:** 24 - cреднее количество обзоров от пользователей, которые поставили больше 48 оценок.

 <div class="alert alert-success">
 <b>👍 Успех:</b> Все верно!
 </div>

 <div class="alert alert-danger">
    <s> <b>😔 Необходимо исправить:</b> Допиши, пожалуйста, итоговый вывод</s>
 </div>

<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Дополнила, см. ниже.</div>

## Итоговый вывод.

В базе данных 1000 книг. Из них 819 книг вышло после 1 января 2000 года. 
Большая часть базы - актуальная, современная литература.

У каждой книги из 1000 есть оценки, отзывы есть для 994 книг. 
Это значит, что в той или иной степени отклик находится для каждой книги.

Максимальное количество отзывов на книгу - 7, минимальное - 0.
Максимальная средняя оценка книги - 5 баллов, минимальная - 2,5 балла.

Наибольшее количество книг толще 50 страниц выпустило издательство Penguin Books. 
Оно может стать одним из ключевых партнеров сервиса.

Автор с самой высокой средней оценкой книг среди книг с 50 и более оценками - J.K. Rowling/Mary GrandPré. 
Новинки от этого автора обязательно должны быть представлены на сервисе.

24 - cреднее количество обзоров от пользователей, которые поставили больше 48 оценок.
Нужно придумать систему поощрения активных пользователей, т.к. их отзывы могут увеличивать продажи.

<div class="alert alert-success">
<b>👍 Успех:</b> Хорошая работа, поздравляю с окончанием курса! Рекомендую еще пройти бесплатный курс от практикума по <a href='https://practicum.yandex.ru/ycloud-datalens/'> DataLens</a>
   Дополнительную информацию по SQL  можно посмотреть и потренироваться <a href='https://sql-ex.ru'> здесь</a>, <a href='https://pgexercises.com'> здесь</a>, <a href='https://sql-academy.org'> здесь</a>, <a href='https://stepik.org/course/63054/info'> здесь</a>, <a href='https://Itresume.ru'> здесь</a>
</div>